st2common/st2common/router.py

```diff
@@ -0,0 +1,386 @@
+        """
+        endpoint, path_vars = self.match(req)
+
+        context = copy.copy(getattr(self, 'mock_context', {}))
+
+        # Handle security
+        if 'security' in endpoint:
+            security = endpoint.get('security')
+        else:
+            security = self.spec.get('security', [])
+
+        if self.auth and security:
+            try:
+                auth_resp = None
+                security_definitions = self.spec.get('securityDefinitions', {})
+                for statement in security:
+                    declaration, options = statement.copy().popitem()
+                    definition = security_definitions[declaration]
+
+                    if definition['type'] == 'apiKey':
+                        if definition['in'] == 'header':
+                            token = req.headers.get(definition['name'])
+                        elif definition['in'] == 'query':
+                            token = req.GET.get(definition['name'])
+                        else:
+                            token = None
+
+                        if token:
+                            if auth_resp:
+                                raise auth_exc.MultipleAuthSourcesError(
+                                    'Only one of Token or API key expected.')
+
+                            auth_func = op_resolver(definition['x-operationId'])
+                            auth_resp = auth_func(token)
+
+                            context['user'] = User.get_by_name(auth_resp.user)
+
+                if 'user' not in context:
+                    raise auth_exc.NoAuthSourceProvidedError('One of Token or API key required.')
+            except (auth_exc.NoAuthSourceProvidedError,
+                    auth_exc.MultipleAuthSourcesError) as e:
+                LOG.error(str(e))
+                return abort_unauthorized(str(e))
+            except auth_exc.TokenNotProvidedError as e:
+                LOG.exception('Token is not provided.')
+                return abort_unauthorized(str(e))
+            except auth_exc.TokenNotFoundError as e:
+                LOG.exception('Token is not found.')
+                return abort_unauthorized(str(e))
+            except auth_exc.TokenExpiredError as e:
+                LOG.exception('Token has expired.')
+                return abort_unauthorized(str(e))
+            except auth_exc.ApiKeyNotProvidedError as e:
+                LOG.exception('API key is not provided.')
+                return abort_unauthorized(str(e))
+            except auth_exc.ApiKeyNotFoundError as e:
+                LOG.exception('API key is not found.')
+                return abort_unauthorized(str(e))
+            except auth_exc.ApiKeyDisabledError as e:
+                LOG.exception('API key is disabled.')
+                return abort_unauthorized(str(e))
+
+        if cfg.CONF.rbac.enable:
+            user_db = context['user']
+
+            permission_type = endpoint.get('x-permissions', None)
```

Comment:
Yes. It only works for most simple RBAC case though. We would have to figure complex cases during RBAC 2.0, but I want us to move in this general direction so we could display rbac requirements in API docs.